### DISCUSSION TITLE
Require reset keypress for System76 Launch

### DISCUSSION THREAD

**Discussion by jackpot51** [fwupd/fwupd#2937] [Q&A]
Right now it resets with a HID command. For security reasons, we want to verify physical access to the keyboard before it can reset to the bootloader. I want to implement this, and I know other plugins do it with some kind of prompt to the user. What is the best example I can look at?

**Comment by hughsie** [comment]
The only thing we have similar are devices like 8bitdo controllers that require you to unplug the controller, hold down a few buttons and replug the controller. There's also the altos ChaosKey, Pinecil and AIAIAI devices which you have to hold down a button on insert to get them into a bootloader that you can query. In the case of the Pinecil there's no USB runtime at all which makes it more of a challenge.

I guess it depends on what you want to do. The easiest thing would be to ask the user to hold down a button while plugging in, and if they do that the "canDownload" bit in the DFU runtime interface gets set to 1, rather than 0. You could obviously implement something else for a custom HID protocol.

There's https://lvfs.readthedocs.io/en/latest/metainfo.html#screenshots to read for that kind of thing; the way that works in fwupd is that the device sets FU_DEVICE_REMOVE_DELAY_USER_REPLUG and FWUPD_DEVICE_FLAG_NEEDS_BOOTLOADER which causes the front end to fwupd to download the URL and show the user an image with a screenshot and caption.

The firmware has to "require" a GUI that can show an image, if it can't work with just a caption: https://lvfs.readthedocs.io/en/latest/metainfo.html#client-features -- the lineart should probably match the style of the existing ones.

Of course, with my security hat on you should probably just require signed firmware images, which means even if the system is compromised you can't flash an unsigned malicious payload onto the keyboard even if you auto-boot it into DFU mode. I personally like the Google Chromebook way of doing this, where you have a signed requirement on firmware, unless you remove the physical screw on the motherboard (voiding your warranty in the process) -- but of course this requires a keyboard controller that's powerful enough to do meaningful crypto, or an ECC/RSA accelerator chip.

**Reply by superm1** [reply]
I agree while plugging it in to press something is probably the best in towards of effort and what the daemon can do.  You can then show a message in the daemon if it wasn't set to tell someone to do that.

If you want it specifically while flashing the other option is to return an update error if you don't hold the button during the entire update or something like that.

**Reply by jackpot51** [reply]
Using QMK as the firmware, and with very little space on the 32 KB firmware image, signing will not be possible. Right now we have a HID command to detach to DFU mode. We would instead want to prompt the user to press Fn+Esc on the keyboard, which would detach to DFU mode.

**Reply by jackpot51** [reply]
Are prompts supported in the fwupd command line?

**Reply by jackpot51** [reply]
Like with `fwupdmgr install`, would the 8bitdo controllers prompt to reboot to flashing mode?

**Reply by superm1** [reply]
Not really - the way that works with 8bitdo is that the plugin returns an "error" and then the user takes action and will re-run the update.  But as @hughsie said we can require that the client show an image or caption.

Changing this architecture to make it truly interactive would mean all clients would need to support this.

**Reply by hughsie** [reply]
> Are prompts supported in the fwupd command line

We only support the before-update caption and the post-install caption text. I think you can probably achieve what you need with the `FWUPD_DEVICE_FLAG_NEEDS_BOOTLOADER` capability.

**Reply by hughsie** [reply]
> Changing this architecture to make it truly interactive

Note: I'm not against doing this if we need to. I'd like some kind of high level UX design work done if that's the case tho, e.g. explaining how the user is going to interact with the various flows. Keeping things simple is sometimes the best design.

**Reply by superm1** [reply]
> Keeping things simple is sometimes the best design

Particularly when you can completely control the firmware behavior, it is certainly the path of least resistance.

**Reply by jackpot51** [reply]
The only issue I see with the method of erroring and then requiring the update to be re-run, is that this is a keyboard and could be the only keyboard on the system. So re-running the update after resetting it to DFU mode where the keyboard is not functional will be difficult.

**Reply by hughsie** [reply]
Rather than the UX being:

1. fwupd update <keyboard>
2. keyboard error not in bootloader mode
3. user presses keys to make it go into bootloader
4. fwupd update <keyboard>

could it be:

1. fwupd update <keyboard>
2. keyboard error as bootloader mode not enabled
3. user presses keys to *enable* bootloader mode
4. fwupd update <keyboard>
5. fwupd detaches to DFU mode, does the update, reattaches to runtime mode

**Reply by jackpot51** [reply]
Yes, we could have a keypress that enables reset to bootloader. That does seem better

**Reply by superm1** [reply]
I guess the only thing you need to keep in mind is whatever key combination you use for this behavior needs to be something that is not normally pressed.  Like if FN-ESC also normally sends an event to the OS (I don't know what you silkscreened the key to do when fn is pressed) it's probably a bad combination.

It should be something solely intercepted by the keyboard MCU and never sent to the OS.

**Reply by hughsie** [reply]
This is when I'm envious of system76 designing the actual hardware.

**Reply by jackpot51** [reply]
Absolutely, it will be handled in the keyboard controller and never sent to the OS.

**Reply by jackpot51** [reply]
We will likely have an LED effect used to indicate the bootloader has been unlocked. This is to make sure users know something unspoofable has happened. No changes need to happen to fwupd except to add a better error in the case it tries to detach before it is unlocked. I will have a PR as soon as I have the firmware changes implemented.

**Reply by hughsie** [reply]
> add a better error

Sounds like a plan, thanks. If we need translations a new FWUPD_ERROR code might be best and then we can map that to an actual string in the UI as a best-effort, falling back to a generic plugin-provided string. If it is a new error you'll need to mark the failure as FWUPD_UPDATE_STATE_FAILED_TRANSIENT I think.

**Comment by superm1** [comment]
There is https://github.com/fwupd/fwupd/discussions/2817 which is useful to look at

**Comment by jackpot51** [comment]
Thanks. This should still be an issue because I will need to implement changes to the plugin to support this.

**Reply by superm1** [reply]
I guess once we're aligned on how you do it, feel free to open another issue and assign to yourself to track.
